src/modules/auth/services/roles.guard.ts

```diff
@@ -0,0 +1,50 @@
+import {
+  CanActivate,
+  ExecutionContext,
+  HttpException,
+  HttpStatus,
+  Inject,
+  Injectable,
+  UnauthorizedException,
+} from '@nestjs/common';
+import { Reflector } from '@nestjs/core';
+import { JwtService } from '@nestjs/jwt';
+import { UserService } from 'src/modules/user/services/user.service';
+import { Services, UserRoles } from 'src/utils/constants';
+import { ROLES_KEY } from './roles-auth.decorator';
+
+@Injectable()
+export class RolesGuard implements CanActivate {
+  constructor(
+    @Inject(Services.USER) private userService: UserService,
+    private jwtService: JwtService,
+    private reflector: Reflector,
+  ) {}
+
+  async canActivate(context: ExecutionContext): Promise<boolean> {
+    try {
+      const requiredRoles = this.reflector.getAllAndOverride<UserRoles[]>(ROLES_KEY, [
+        context.getHandler(),
+        context.getClass(),
+      ]);
+      if (!requiredRoles) {
+        return true;
+      }
+      const req = context.switchToHttp().getRequest();
+      const authHeader = req.headers.authorization;
+      const bearer = authHeader.split(' ')[0];
+      const token = authHeader.split(' ')[1];
```

Comment:
you have it in JwtAuthGuard why you need it here again?

Reply:
I thought that we can use Roles guards combined with Jwt verifying, so that we can use one Role Guard in endpoints instead of two (Role and Jwt). And also in FE we will put JWT token in headers and then verify it and take data from it by using Role Guard instead of putting user data in request body/params. But I don't know how it will be better. So I decided to leave Jwt guard and Role guard in the project, maybe to use two of them in the future.
I think I need your opinion - how do you think it would be better?

Reply:
I dont see when 2 are needed. what is difference between them?